File: DS/SORTING/Quick-sort.py
a = [8,2,5,6,8,1,3]

def partition(a,start,end):
    p = a[end]
    pindex = start
    for i in range(start, end):
        if a[i]<=p:
            t = a[i]
            a[i] = a[pindex]
            a[pindex] = t
            pindex+=1
    swap(a[pindex], a[end])
    t = a[pindex]
    a[pindex] = p
    a[end] = t
    return pindex

def quicksort(a,start,end):
    if start<end:
        pindex = partition(a,start,end)
        quicksort(a,start,pindex-1)
        quicksort(a,pindex+1,end)
        
quicksort(a,0,len(a)-1)
print(a)
    

File: DS/SORTING/QuickSort.java
class QuickSort {

    public static void main(String... args){
        int[] a = {8,5,8,7,2,3,9,1};
        quicksort(a,0, a.length-1);
        System.out.println(java.util.Arrays.toString(a));
    }

    static void quicksort(int[] a, int start, int end){
        if(start<end){
            int pindex = partition(a,start,end);
            quicksort(a, start, pindex-1);
            quicksort(a, pindex+1, end);
        }
    }

    static int partition(int[] a, int start, int end){
        int pivot = a[end];
        int pindex = start;
        for(int i = start; i<end; i++){
            if(a[i]<=pivot){
                int t = a[pindex];
                a[pindex] = a[i];
                a[i] = t;
                pindex++;
            }
        }
        int t = a[pindex];
        a[pindex] = a[end];
        a[end] = t;
        return pindex;
    }
}